packages/zipkin-instrumentation-request-promise/src/promise.js

```diff
@@ -0,0 +1,265 @@
+/* eslint-disable no-underscore-dangle, no-use-before-define, no-param-reassign*/
+import asap from 'asap/raw';
+
+const noop = () => {};
+
+// States:
```

Comment:
did you write all the code in this file yourself? If so, can you put some cursory notes about what the classes and functions do. If not, can you cite where the code is from? If in any case docs can be simplified by referring to similar libraries, just mention the library it is similar to.

Reply:
**Recap**: The reason `request` break the tracing is due to the promise losing or not restoring the original execution context. You can't override the default Node promise callback execution behavior, I won't get into the explanation it's a long story. So I created a promise capable to restore the original context before callback execution.
I wasn't comfortable writing a promise myself from scratch so I found this library [then/promise](https://github.com/then/promise/blob/master/src/core.js) then redesign to create my `promise.js` (I didn't violate the code [license](https://github.com/then/promise/blob/master/LICENSE)). The diff between these files will be the code I added.

**Side note:** The other instrumented libraries like `zipkin-instrumention-redis` using promises many have the same issue.

Reply:
thanks for the notes and the side note. Can you put the recap at the top of the file in some form?

That's it from me unless others have things to say.